_episodes/07-b-numpy.md

```diff
@@ -0,0 +1,280 @@
+
+## NumPy Arrays vs. Python Lists
+
+Previously, you have worked with the built-in types of `lists`. NumPy arrays seem similar, but offer some distinct advantages. We have also worked with Numpy arrays, but we have not taken advantage of their unique features. 
+
+Numpy arrays take up less space, are faster, and have more mathematical operations associated with them. However, unlike lists, they elements all have to be the same type. This is why we had to read in the file from the Tabular Data lesson using the `dtype` argument.
+
+There are also differences in how lists and numpy arrays behave. Let's look at some of these.
+
+To use the numpy library, we have to import it. When `numpy` is imported, it is often shortened to `np`.
+
+~~~
+import numpy as np
+
+# Create some lists
+r1 = [0.0, 0.1, 0.0]
+r2 = [0.0, 0.0, 0.0]
+
+# Create some numpy arrays
+r1_array = np.array(r1)
+r2_array = np.array(r2)
+~~~
+{: .language-python}
+
+In this code block, we've created two lists (`r1` and `r2`). We then created numpy array versions of these lists.
+
+When we print `r1` and `r1_array`, the results look very similar.
+
+~~~
+print(r1)
+print(r1_array)
+~~~
+{: .language-python}
+
+~~~
+[0.0, 0.1, 0.0]
+[0.  0.1 0. ]
+~~~
+{: .output}
+
+However, one way numpy arrays and lists vary significantly is that you can easily perform element-wise operations on numpy arrays without loops. **This is advantageous because loops in Python are slow. However, `numpy` functions are optimized to run quickly.** If possible, you should avoid loops when working with numpy arrays.
+
+You can add two arrays together, multiply arrays by scalars, or do element-wise multiplcation of arrays.
+
+For example, you can multiply two numpy arrays to get their element-wise product. This means that given two vectors `a = np.array([a0, a1, a2])` and `b = np.array([b0, b1, b2])`, `a * b = [a0*b0, a1*a1, a2*b2]`.
+
+In contrast, if `a` and `b` were lists, you would get an error. 
+
+> ## Check your understanding
+> What does the following code result in?
+>
+> ~~~
+> a1 = np.array([2, 1, 0])
+> a2 = np.array([1, 3, 5])
+> 
+> print(a1 * a2)
+> ~~~
+> {: .language-python}
+> 
+> What happens if a1 and a2 are lists?
+>
+>> ## Answer
+>> 
+>> The code block results in 
+>> ~~~
+>> [2 3 0]
+>> ~~~
+>> {: .output}
+>> The first element is `a1[0]*a2[0]`, the second element is `a1[1]*a2[1]`, and the third element is `a1[2]*a2[2]`.
+>>
+>> If a1 and a2 were lists instead, you would get a `TypeError`. `TypeError: can't multiply sequence by non-int of type 'list'`.
+>>
+> {: .solution}
+{: .challenge}
+
+## Returning to Tabular Data
+
+Let's consider our tabular data.
+
+~~~
+import os
+import numpy
+
+distance_file = os.path.join('data', 'distance_data_headers.csv')
+distances = numpy.genfromtxt(fname=distance_file, delimiter=',', dtype='unicode')
+headers = distances[0]
+data = distances[1:]
+data = data.astype(numpy.float)
+~~~
+{: .language-python}
+
```

Comment:
I would move all of this to an exercise, stating what you want the name of each variable to be to make sure it's consistent with below.  In a real teaching situation, I would almost always have the students do this themselves to review what they already learned.